control/autopilot/include/autopilot/autopilot_inl.h

```diff
@@ -1147,20 +1222,57 @@ quadrotor_common::ControlCommand AutoPilot::executeTrajectory(
       *trajectory_execution_left_duration = ros::Duration(0.0);
       *trajectories_left_in_queue = 0;
       trajectory_queue_.pop_front();
-      setAutoPilotState(States::HOVER);
-      return base_controller_.run(state_estimate, reference_state_,
+      setAutoPilotStateForced(States::HOVER);
+
+      reference_trajectory_ = quadrotor_common::Trajectory();
+      reference_trajectory_.trajectory_type = 
+        quadrotor_common::Trajectory::TrajectoryType::GENERAL;
+      reference_trajectory_.points.push_back(reference_state_);
+      return base_controller_.run(state_estimate, reference_trajectory_,
                                   base_controller_params_);
     }
     else
     {
       time_start_trajectory_execution_ +=
-          trajectory_queue_.front().points.back().time_from_start;
+        trajectory_queue_.front().points.back().time_from_start;
       trajectory_queue_.pop_front();
     }
   }
 
   const ros::Duration dt = time_now - time_start_trajectory_execution_;
   reference_state_ = trajectory_queue_.front().getStateAtTime(dt);
+  reference_trajectory_ = quadrotor_common::Trajectory();
+  reference_trajectory_.trajectory_type = 
+    quadrotor_common::Trajectory::TrajectoryType::GENERAL;
+
+  std::list<quadrotor_common::TrajectoryPoint>::const_iterator it_points;
+  std::list<quadrotor_common::Trajectory>::const_iterator it_trajectories;
+  bool lookahead_reached(false);
```

Comment:
- could you comment the code, basic idea behind the heuristic (i'm lazy but so will be the next person working on it)
- should it not be a separate function like mpc?
- could we use specialization for these functions that change a lot ? I mean a definition for execute(mpc controller templated) and for execute(normal controller) ... Idk if it's possible but I think so. [](http://en.cppreference.com/w/cpp/language/template_specialization
)

Reply:
3rd bullet: nope... if so we need to specify the template type in here... then we need the type at compile time.
but this is what we want to avoid, to be able to compile rpg_quadrotor_control without rpg_mpc.
I will think about making it more readable and understandable.

Reply:
I will look at this in more detail later then

Reply:
@foehnph Agreed for the 3rd bullet, you are right!

Reply:
Ok temporary solution:
I commented the code there to make it understandable what it does.
I think this should be a separate class, like `autopilot::Tracker`, the trajectory queue could be passed as reference and it can have time-based tracking and spacial tracking, as well as restrictions to forward-only tracking (can speed up iterations massively).
Additionally it could also handle the trajectory queue if wanted, but I guess it's easier to leave that in the autopilot.